### PR TITLE
Impelment the module's core features

### PR DESCRIPTION
**Pull request by deleepa**
This PR contains the implementation of the core features for this module.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -0,0 +1,127 @@
+'use strict';
```
Comment (rehrumesh) [inline index.js:1]: When it comes to JS, better to use 2 spaces as a tab. Refer Airbnb JS Style guide. We are using that guide inside 99X JS projects.

File: index.js [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+let AwsCustomAuthoriserToolkit = {};
+
+AwsCustomAuthoriserToolkit.buildPolicyFromIam = (jwtIdToken, callback) => {
+
+    let authToken = jwtIdToken;
+    let iam = new aws.IAM();
+    let decodedJwt = jwt.decode(authToken);
+
+    console.log(decodedJwt);
```
Comment (rehrumesh) [inline index.js:16]: If the console log is unnecessary, please remove it. JWT may consists with sensitive data which will ultimately recorded in the CloudWatch logs. May be add a verbose mode for logging throughout the project So the logging for sensitive data will happen only requested.

File: index.js [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+    let roleNames = _.map(decodedJwt["cognito:roles"], function (arn) {
+        return _.last(arn.split('/'));
+    });
+
+    let getPolicyNames = function (roleName, callback) {
+        iam.listRolePolicies({ RoleName: roleName }, function (err, data) {
+            var policyNames = data["PolicyNames"];
```
Comment (rehrumesh) [inline index.js:24]: Any particular reason to use `var`? If not we can replace it with `let`

File: index.js [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    let iam = new aws.IAM();
+    let decodedJwt = jwt.decode(authToken);
+
+    console.log(decodedJwt);
+
+    let roleNames = _.map(decodedJwt["cognito:roles"], function (arn) {
+        return _.last(arn.split('/'));
+    });
+
+    let getPolicyNames = function (roleName, callback) {
```
Comment (rehrumesh) [inline index.js:22]: Consider using named function expression. Applicable to every function declaration.

File: index.js [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+        console.log(statement);
+        return policyDocument;
+    };
+
+    let iamPolicyDocumentRequests = _.map(roleNames, function (roleName) {
+        return function (callback) {
+            async.waterfall([
+                function (callback) {
+                    console.log('get roleName', roleName);
```
Comment (rehrumesh) [inline index.js:61]: Consider removing `console.log`

File: index.js [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+                    getPolicyNames(roleName, callback);
+                }, function (policyNames, callback) {
+                    let policyDocumentRequests = getPolicyDocumentRequets(policyNames, roleName, callback);
+                    async.parallel(policyDocumentRequests, function (err, policyDocuments) {
+                        console.log('get documents', policyDocuments);
```
Comment (rehrumesh) [inline index.js:66]: Consider removing `console.log`

File: index.js [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+        returnPolicy.principalId = decodedJwt["sub"];
+        returnPolicy.policyDocument = policyDocument;
+
+        callback(null, returnPolicy);
+    });
+};
+
+AwsCustomAuthoriserToolkit.validateCognitoIdToken = (jwtToken, jwkPem, iss, callback) => {
+    
+    var decodedJwt = jwt.decode(jwtToken, {complete: true});
```
Comment (rehrumesh) [inline index.js:89]: consider using `let` instead of `var`

File: index.js [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+    
+    //Fail if the token is not jwt
+    if (!decodedJwt) {
+        callback("Not a valid JWT token");
+        return;
+    }
+
+    //Fail if token is not from your User Pool
+    if (decodedJwt.payload.iss != iss) {
```
Comment (rehrumesh) [inline index.js:98]: `!==` instead of `!=`

File: index.js [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+        callback("invalid issuer");
+        return;
+    }
+
+    //Reject the jwt if it's not an 'Access Token'
+    if (!(decodedJwt.payload.token_use == 'id' || decodedJwt.payload.token_use == 'access')) {
```
Comment (rehrumesh) [inline index.js:104]: `===` instead of `==`

File: index.js [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+        callback("token_use is invalid");
+        return;
+    }
+
+    //Get the kid from the token and retrieve corresponding PEM
+    var kid = decodedJwt.header.kid;
```
Comment (rehrumesh) [inline index.js:110]: Consider `let` over `var`

File: index.js [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+    var pem = jwkPem[kid];
```
Comment (rehrumesh) [inline index.js:111]: Consider `let` over `var`

**Review by rehrumesh — CHANGES_REQUESTED** [review]
Please look into the inline comments

File: index.js [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+
+        let statement = _.flatMap(policyDocuments, function (policyDocument) {
+            var policyDocumentObj = JSON.parse(policyDocument);
+            return policyDocumentObj.Statement
+        });
+
+        policyDocument.Version = "2012-10-17";
+        policyDocument.Statement = statement;
+
+        console.log(statement);
```
Comment (rehrumesh) [inline index.js:53]: Consider removing console log.

**Comment by AshanFernando** [comment]
@rehrumesh I will fixed the mentioned points

**Pull request merged**